lambeq/rewrite/base.py

```diff
@@ -393,3 +395,82 @@ def _ar(self, box: Box) -> Diagram:
 
     def _ob(self, ob: Ty) -> Ty:
         return ob
+
+
+class UnknownWordsRewriteRule(RewriteRule):
+    """A rewrite rule for unknown words.
+
+    This rule matches the word in the given list of words of and replaces
+    with UNK and when passed a diagram, replaces all the boxes
+    containing an unknown word with an UNK box corresponding to the same
+    pregroup type.
+    """
+    def __init__(self,
+                 unknown_words: Container[str] | None = None) -> None:
+        """Instantiate a UnknownWordsRewriteRule.
+
+        Parameters
+        ----------
+        words : container of str, optional
+            A list of words to be rewritten by this rule. If a box does
+            not have one of these words, it will not be rewritten with
+            UNK.
+        """
+        self.unknown_words = [] if unknown_words is None else unknown_words
+
+    def matches(self, box: Box) -> bool:
+        if box.name in self.unknown_words:
+            return True
+        return False
+
+    def rewrite(self, box: Box) -> Diagram:
+        return type(box)('UNK',  dom=box.dom, cod=box.cod)
+
+
+class HandleUnknownWords:
+    """Handle unknown words in diagrams using a minimum frequency. Words
+    that appear less than `min_freq` times in the training data are
+    replaced with UNK boxes. This rule is used in conjunction with
+    :py:class:`UnknownWordsRewriteRule` to replace unknown words in
+    diagrams.
+    """
+    def __init__(self, min_freq: int = 1) -> None:
+        """Instantiate a HandleUnknownWords rule.
+
+        Parameters
+        ----------
+        min_freq : int, optional
+            The minimum frequency of a word to be considered known.
+        """
+        self.min_freq = min_freq
+        self.unknown_words = set()
+
+    def train(self, diagrams: List[Diagram],
+              strings: Optional[List[str]] = None):
+        word_counts = Counter()
+        if strings is not None:
+            for string in strings:
+                word_counts.update(string.split())
+        for diagram in diagrams:
+            for box in diagram.boxes:
+                if isinstance(box, Word):
+                    word_counts[box.name] += 1
+        self.unknown_words = set(word
+                                 for word, count in word_counts.items()
+                                 if count < self.min_freq)
+
+    def test(self, diagrams: List[Diagram],
+             unknown_words: List[str]) -> List[Diagram]:
+        rule = UnknownWordsRewriteRule(unknown_words=unknown_words)
+        rewriter = Rewriter([rule])
+        rewritten_diagram = rewriter(diagrams)
```

Comment:
A rewriter acts on a single diagram, rather than a list of diagrams, so currently if a list of diagrams is passed both the `train` and `test` methods throw an error.